### PR TITLE
refactor(IscDhclient): discover DHCP leases at distro-provided location

### PR DESCRIPTION
**Pull request by phsm**
## Proposed Commit Message

```
refactor(IscDhclient): discover DHCP leases at distro-provided location

Use leases directory and lease file provided by Distro object
instead of cycling through all possible locations.
```

## Additional Context

IscDhclient class seems to be only used by CloudStack datasource.

## Checklist

- [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/development/index.html)
- [x] I have updated or added any [unit tests](https://cloudinit.readthedocs.io/en/latest/development/testing.html) accordingly
- [x] I have updated or added any [documentation](https://cloudinit.readthedocs.io/en/latest/development/contribute_docs.html) accordingly

## Merge type

- [x] Squash merge using "Proposed Commit Message"
- [ ] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

### CODE REVIEW

**Review by igalic — APPROVED** [review]
This looks good to me!

This part of the commit message,

> This change adds allows discovering such lease files.

seems like it should be either "adds" or "allows", not both.

**Comment by phsm** [comment]
> This looks good to me!
> 
> This part of the commit message,
> 
> > This change adds allows discovering such lease files.
> 
> seems like it should be either "adds" or "allows", not both.

Thanks, fixed the typo in the commit message.

File: cloudinit/net/dhcp.py
Comment (holmanb) [inline cloudinit/net/dhcp.py:476]: I'm on the fence about whether to make this a (free)bsd check before adding. On one hand, I check a couple of linuxen, and some have the directory, others don't. Why check this directory on those distros? On the other hand, there were few files in that directory and only a single `listdir` called, so it shouldn't be "slow".

This function is one that I've previously considered breaking out into `distros/` anyways, since I think `/var/lib/dhcp/` and `/var/lib/dhclient/` are distro-specific iirc.

Thoughts?

**Review by holmanb — COMMENTED** [review]
Thanks for this @phsm!

Nice addition and tests, I'm leaning towards making that a (free)bsd check prior to adding to the list.

Justification below.

File: cloudinit/net/dhcp.py
Comment (phsm) [reply cloudinit/net/dhcp.py:476]: I briefly checked the `distros/` directory: seems its purpose is to isolate distro-specific code in there. So this means that the cloud-init has a reliable way to detect which os/distro it is running on? 

If yes, then its indeed better to implement this function per-distro.
Perhaps implement a method in `distros.Distro` that uses a class attribute [such as this](https://github.com/canonical/cloud-init/blob/main/cloudinit/distros/debian.py#L34) and override that class attribute per child distro class.

Comment (holmanb) [reply cloudinit/net/dhcp.py:476]: > Perhaps implement a method in distros.Distro that uses a class attribute with the dhclient directory [such as this](https://github.com/canonical/cloud-init/blob/main/cloudinit/distros/debian.py#L34) and override that class attribute per child distro class.

@phsm yes, this sounds like the right direction. I think we have a reference to the distro instance higher up that we'll just need to pass down the stack as an argument to access it in this function.

Comment (phsm) [reply cloudinit/net/dhcp.py]: I just made another commit with an example how I see it done:
- main Distro class has two attributes: dhclient_lease_directory, dhclient_lease_file_regex defining the lease directory for lease files in a distro, and a regex helping to find those lease files in there.
- every class of a distro that uses dhclient has those 2 attributes defined.
- Cloudstack's get_vr_address() now detects a distro and supplies dhcp.IscDhclient.get_latest_lease() with the lease_dir an a regex.
- there is no need for get_dhclient_d() anymore
- Although get_latest_lease() function signature is changed, I don't see it being used anywhere except Cloudstack DS so it shouldn't break anything.

The purpose of this commit is to gather your feedback if i'm working in the right direction. 
Also, do I detect the distro correctly here?

File: cloudinit/sources/DataSourceCloudStack.py
Comment (holmanb) [inline cloudinit/sources/DataSourceCloudStack.py:286]: Driveby comment:  Why call Init()? `__init__()` already got an instance of distro, lets just save a reference to it in an instance variable and use that rather than calling `Init()`.

Comment (phsm) [reply cloudinit/sources/DataSourceCloudStack.py]: Indeed, I noticed that the Distro object is already passed into `sources.DataSource.__init__()` as an argument. Thus, no need to obtain the Distro object once more.
Switched to using that. 

Does the rest look OK so I can implement it for the other distros?

File: cloudinit/distros/rhel.py
Comment (igalic) [inline cloudinit/distros/rhel.py:35]: where is the equivalent for this in other distros? particularly freebsd

**Review by igalic — COMMENTED** [review]
❓

File: cloudinit/distros/rhel.py
Comment (phsm) [reply cloudinit/distros/rhel.py:35]: Hi @igalic,
As I mentioned earlier, this commit is intended as a WIP, a draft to show an idea how I intend to implement it. I just picked RHEL distro as an example. 

Once I get Brett's approval on this course, I'll implement it for every distro, FreeBSD included.

File: cloudinit/net/dhcp.py
Comment (phsm) [reply cloudinit/net/dhcp.py]: I just updated the PR with the following:
1. Updated the distros files with dhclient lease director/regex for distros that support DHClient. I installed all of them on a virtual machine to verify the paths:

- CentOS: updated centos.py to contain dhclient parameters for Centos 7. I assume no one uses CentOS > 7 on servers because they're EOL.
- RHEL: updated rhel.py with dhclient parameters. To use dhclient in RHEL, `dhcp=dhclient` parameter must be set in NetworkManager config.
- AlmaLinux: identical to RHEL.
- Rocky: identical to RHEL.
- Debian: updated debian.py with dhclient parameters.
- Alpine: doesn't use dhclient, uses udhcpc.
- Amazon Linux: updated amazon.py with dhclient parameters.
- CloudLinux: identical to RHEL.
- EuroLinux: identical to RHEL.
- Fedora: identical to RHEL
- FreeBSD: updated freebsd.py with dhclient parameters.
- Gentoo: doesn't use dhcpcd, uses dhcpcd.
- Mariner: uses systemd-networkd.
- Miracle: identical to RHEL
- Open Cloud OS: I couldn't even download it an ISO image to test. I assume its identical to RHEL.
- OpenEuler: identical to RHEL
- OpenSUSE: doesn't use dhclient, uses wicked.
- SLES: I don't have a copy of SLES, I assume it also uses wicked.
- Virtuozzo: I don't have a copy of it, I assume its identical to RHEL.
- Tencentos: didn't test, assume its identical to RHEL. 

2. Tests: 
- Added tests for finding a lease in different distros that can use dhclient.
- Removed `TestGetLatestLease` testcase from `test_cloudstack.py` in its entirety: Cloudstack tests shouldn't test obtaining DHCP lease, its already covered in `test_dhcp.py`

**Comment by phsm** [comment]
Sorry for commit storm, I was struggling with the linters and the tests.

I also changed the PR description as its not just fixing Dhclient at FreeBSD anymore.

File: cloudinit/sources/DataSourceCloudStack.py
Comment (igalic) [inline cloudinit/sources/DataSourceCloudStack.py:299]: I'm not sure this should happen at all

File: cloudinit/sources/DataSourceCloudStack.py
Comment (igalic) [inline cloudinit/sources/DataSourceCloudStack.py:300]: and if it did, i don't see how a debug log is useful.

remove the if, and the else, let it crash loudly and clearly

File: cloudinit/sources/DataSourceCloudStack.py
Comment (igalic) [inline cloudinit/sources/DataSourceCloudStack.py:114]: please get rid of this if, I think it's misleading.
this reads as: this could sensibly happen! rather than: if this happens, nothing here should work

**Review by igalic — CHANGES_REQUESTED** [review]
this looks good, here's some requests for changes

File: cloudinit/sources/DataSourceCloudStack.py
Comment (phsm) [reply cloudinit/sources/DataSourceCloudStack.py:299]: Hi @igalic, 

Thank you for your comments. I believe all three your comments can be grouped here together in one notion: "explicit is better than implicit, better return an error than continuing with undefined behavior".

I normally would agree with it but there is one more last resort down the code: look for Cloudstack Metadata server at server's default gateway. So even if something is wrong with the DHCP client code, there's still a chance for cloud-init to get its hands on metadata.

I know its hard to discuss the logical flow without knowing whats going on underneath. That's why I explained how Cloudstack networks work in detail in my previous PR: https://github.com/canonical/cloud-init/pull/4433#discussion_r1366978013

Basically the reason behind not raising an exception here is to allow the code to use the last resort, the default gateway address.

File: cloudinit/sources/DataSourceCloudStack.py
Comment (phsm) [reply cloudinit/sources/DataSourceCloudStack.py:114]: This method was added in my previous PR: https://github.com/canonical/cloud-init/pull/4433
You can follow the discussion there (its quite long, though). 

In short: the virtual machine FQDN is assembled by concatenating isntance name + '.' + domain_name_given_by_dhcp.

This method is only used at get_hostname() method that tries to make an FQDN and if it can't obtain DHCP domain name for any reason, it uses the instance name as FQDN (the behavior previous to my PR)

Comment (phsm) [reply cloudinit/sources/DataSourceCloudStack.py:114]: Oh please disregard my previous comment, I thought you commented similar code in get_vr_address() method.

Well, I of course can let it crash here. But what to do with the get_hostname() method that uses this? The current behavior is to try assembling an FQDN and if that fails, return the instance name as FQDN.

File: cloudinit/sources/DataSourceCloudStack.py
Comment (igalic) [reply cloudinit/sources/DataSourceCloudStack.py:299]: I don't doubt that you understand the CloudStack code way better than I do, so let's reformulate my change request into two clear and concise questions:

1. under which circumstances can this `distro` object be `None`?
2. why is `distro=None` a desirable state for a `DataSourceCloudStack` instance to be in?

Comment (phsm) [reply cloudinit/sources/DataSourceCloudStack.py:299]: 1. I don't have an answer to this question. I found that None is passed into DataSourceCloudStack.__init__() during tests so I thought it is better to include a check for it in case None is somehow passed in.
I would assume that distro=None should normally never happen.

2. It is not a desirable state.

I guess if the crash during distro=None is acceptable, I could assume that distro is always provided, and adjust the tests to be initialized with `distros.ubuntu.Distro`  instead of None.

Comment (phsm) [reply cloudinit/sources/DataSourceCloudStack.py]: Updated the PR: removed the distro=None checks, as suggested.

File: cloudinit/sources/DataSourceCloudStack.py
Comment (phsm) [reply cloudinit/sources/DataSourceCloudStack.py]: Updated the PR: removed the distro=None checks, as suggested.

File: cloudinit/sources/DataSourceCloudStack.py
Comment (phsm) [reply cloudinit/sources/DataSourceCloudStack.py]: Updated the PR: removed the distro=None checks, as suggested.

File: cloudinit/sources/DataSourceCloudStack.py
Comment (igalic) [reply cloudinit/sources/DataSourceCloudStack.py]: if it's only happening during tests we should just fix the tests

**Review by igalic — APPROVED** [review]
I like this.
thank you for this monumental effort

**Review by holmanb — CHANGES_REQUESTED** [review]
@phsm nice work here - thanks for putting in this effort.

I left a couple of comments that I think we need to address.

File: cloudinit/distros/rhel.py
Comment (phsm) [reply cloudinit/distros/rhel.py:32]: Implemented:
- if the distro-provided dir and regex yield nothing, then try to use the fallback dir and regex. Regex is specifically made to be quite loose (search for any file ending with `.lease` or `.leases`
- Tests for that behavior

File: cloudinit/distros/centos.py
Comment (phsm) [reply cloudinit/distros/centos.py]: I checked Centos9-stream. As I expected, it is identical to RHEL.
Removed dir and regex definitions under Centos distro, and the test for centos7 case.

File: cloudinit/net/dhcp.py
Comment (holmanb) [inline cloudinit/net/dhcp.py:33]: I think you meant `.+\.lease*$` to match `.lease` or `.leases`, right?

Also we typically avoid inline comments in this codebase as a style rule, fyi.

**Review by holmanb — CHANGES_REQUESTED** [review]
Nice work - getting close! A couple more changes requested, I'm testing this on Ec2 now.

File: cloudinit/net/dhcp.py
Comment (phsm) [reply cloudinit/net/dhcp.py]: well, `s*` means 0 or more occurrences of character `s` ;) So it will match '.lease', '.leases', '.leasessss'.

I moved the inline comment.

Comment (igalic) [reply cloudinit/net/dhcp.py]: shouldn't it just be `\.leases?$` ?

File: cloudinit/distros/rhel.py
Comment (holmanb) [reply cloudinit/distros/rhel.py:32]: @phsm Just looking at this change now, I realize that you made `/var/lib/dhclient/` default for all distros.

From what I can tell, the `dhclient` subdirectory is a feature of RHEL derivatives and Arch. Sorry I missed this before.

The following distros use `/var/lib/dhcp/`: including Alpine, Gentoo (and derivatives I assume), Suse (and derivatives), Debian, etc.

No clue what Photon or Mariner use.

With the change as it stands, `/var/lib/dhclient` will be the only directory checked for distros that do not inherit from FreeBSD / RHEL / Debian, so I think that we need to account for those mentioned above.

@igalic is `/var/lib/dhclient` is a reasonable directory for the non-FreeBSD BSDs?

File: cloudinit/net/dhcp.py
Comment (phsm) [reply cloudinit/net/dhcp.py]: Indeed, I totally forgot about the `?` quantifier. Thanks, fixed that.

File: cloudinit/distros/rhel.py
Comment (igalic) [reply cloudinit/distros/rhel.py:32]: I'll need to check.

Comment (igalic) [reply cloudinit/distros/rhel.py:32]: OpenBSD puts it in the same place as FreeBSD:
https://man.openbsd.org/dhclient.8#FILES

NetBSD had switched to `dhcpcd` since 9.0. before that, they put it in the same place, https://man.netbsd.org/NetBSD-8.2/dhclient.8#FILES but with a different name

File: cloudinit/net/dhcp.py
Comment (holmanb) [reply cloudinit/net/dhcp.py]: Thanks @phsm for the update. I think that there is a missing detail here.

> doesn't use dhclient

Use of wicked / networkd / networkmanager / etc doesn't necessarily mean that dhclient isn't used. In fact, cloud-init itself uses dhclient during the [local](https://cloudinit.readthedocs.io/en/latest/explanation/boot.html#local) stage. Cloud-init sets up an ephemeral network so that it can reach out to the IMDS to get configuration. This allows cloud-init to write the networking configuration locally *before* the distro's built-in network manager is even started.

You've done some good work on this refactor, but I think that by default all distros should probably be checking for the isc-dhclient path, at least until `udhcpc` is fully featured for alpine, etc (with classless static routes and unknown dhcp option support) and dhcpcd replaces isc-dhclient on other distros.

**Review by holmanb — APPROVED** [review]
Great work @phsm, thanks for putting this together.

Looks good to me, so I'll merge.

If I understand correctly, cloudstack might also benefit from parsing the default lease that cloud-init gets during the ephemeral stage. I'll probably include that as part of the work I'm doing with dhcpcd.

**Pull request merged**
